File: python/test/tools/chimera_test.py
#
# Project NuriKit - Copyright 2026 SNU Compbio Lab.
# SPDX-License-Identifier: Apache-2.0
#

from pathlib import Path

import numpy as np
import pytest

from nuri.fmt.pdb import Model, read_models
from nuri.tools import chimera


def _read_calphas(model: Model):
    idxs = [
        i
        for res in model.residues
        for i in res.atom_idxs
        if model.atoms[i].name == "CA"
    ]
    return model.major_conf[idxs]


@pytest.fixture()
def query(test_data: Path):
    model = read_models(test_data / "fixed_ref1.pdb")[0]
    return _read_calphas(model)


@pytest.fixture()
def templ(test_data: Path):
    model = read_models(test_data / "fixed_ref2.pdb")[0]
    return _read_calphas(model)


def test_match_maker(query: np.ndarray, templ: np.ndarray):
    mm = chimera.match_maker(query, templ)

    ref_xform = np.array(
        [
            [-0.14150449, 0.67616094, -0.72303725, 16.25602271],
            [-0.26213529, -0.72990790, -0.63128405, 38.51123346],
            [-0.95460021, 0.10020405, 0.28053089, 17.46343176],
            [0, 0, 0, 1],
        ]
    )
    # fmt: off
    ref_sel = np.array(
        [
            16, 40, 78, 80, 99, 102, 103, 104, 105, 106, 108, 109, 110, 111,
            112, 113, 114, 115, 116, 117, 118, 119, 121, 122, 123, 124, 125,
            126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138,
            139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151,
            153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165,
            166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178,
            179, 180, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193,
            194, 195, 196,
        ]
    )
    # fmt: on

    np.testing.assert_allclose(mm.transform, ref_xform, rtol=1e-6)
    assert mm.aligned_rmsd == pytest.approx(0.698, abs=1e-3)

    mm.selected.sort()
    np.testing.assert_array_equal(mm.selected, ref_sel)


def test_match_maker_errors(query: np.ndarray, templ: np.ndarray):
    with pytest.raises(ValueError, match="have different number of points"):
        chimera.match_maker(query[:3], templ)

    with pytest.raises(ValueError, match="cutoff must be positive"):
        chimera.match_maker(query, templ, -1)

    with pytest.raises(
        ValueError,
        match="global_ratio must be between 0 and 1",
    ):
        chimera.match_maker(query, templ, global_ratio=1.5)

    with pytest.raises(
        ValueError,
        match="viol_ratio must be between 0 and 1",
    ):
        chimera.match_maker(query, templ, viol_ratio=1.5)

    query[0] = np.nan
    with pytest.raises(RuntimeError, match="alignment failed"):
        chimera.match_maker(query, templ)

File: test/tools/chimera_test.cpp
//
// Project NuriKit - Copyright 2026 SNU Compbio Lab.
// SPDX-License-Identifier: Apache-2.0
//

#include "nuri/tools/chimera.h"

#include <filesystem>
#include <fstream>

#include <Eigen/Dense>
#include <eigen3/Eigen/src/Core/Matrix.h>

#include <gtest/gtest.h>

#include "nuri/eigen_config.h"
#include "fmt_test_common.h"
#include "test_utils.h"
#include "nuri/fmt/pdb.h"

namespace nuri {
namespace internal {
namespace {
namespace fs = std::filesystem;

Matrix3Xd read_first_calphas(const fs::path &path) {
  std::ifstream ifs(path);
  PDBReader reader(ifs);
  PDBModel model = read_pdb_model(reader.next());

  std::vector<int> calphas;
  for (const PDBResidue &res: model.residues()) {
    for (int i: res.atom_idxs()) {
      const PDBAtom &atom = model.atoms()[i];
      if (atom.name() == "CA") {
        calphas.push_back(i);
        break;
      }
    }
  }

  return model.major_conf()(E::all, calphas);
}

TEST(ChimeraTest, MatchMaker) {
  Matrix3Xd query = read_first_calphas(test_data("fixed_ref1.pdb")),
            templ = read_first_calphas(test_data("fixed_ref2.pdb"));

  Isometry3d ref_xform;
  ref_xform.linear() << -0.14150449, 0.67616094, -0.72303725,  //
      -0.26213529, -0.72990790, -0.63128405,                   //
      -0.95460021, 0.10020405, 0.28053089;
  ref_xform.translation() << 16.25602271, 38.51123346, 17.46343176;

  ArrayXi ref_sel(95);
  ref_sel << 16, 40, 78, 80, 99, 102, 103, 104, 105, 106, 108, 109, 110, 111,
      112, 113, 114, 115, 116, 117, 118, 119, 121, 122, 123, 124, 125, 126, 127,
      128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142,
      143, 144, 145, 146, 147, 148, 149, 150, 151, 153, 154, 155, 156, 157, 158,
      159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173,
      174, 175, 176, 177, 178, 179, 180, 183, 184, 185, 186, 187, 188, 189, 190,
      191, 192, 193, 194, 195, 196;

  auto mm = match_maker(query, templ);
  ASSERT_NEAR(std::sqrt(mm.msd), 0.698, 1e-3);

  ASSERT_EQ(mm.sel.size(), 95);
  absl::c_sort(mm.sel);
  for (int i = 0; i < 95; ++i)
    EXPECT_EQ(mm.sel[i], ref_sel[i]) << i;

  NURI_EXPECT_EIGEN_EQ_TOL(mm.xform.linear(), ref_xform.linear(), 1e-6);
  NURI_EXPECT_EIGEN_EQ_TOL(mm.xform.translation(), ref_xform.translation(),
                           1e-5);

  Matrix3Xd aligned(3, query.cols());
  inplace_transform(aligned, mm.xform, query);
  double msd = (aligned - templ).colwise().squaredNorm().mean();
  EXPECT_NEAR(std::sqrt(msd), 7.472, 1e-3);
}
}  // namespace
}  // namespace internal
}  // namespace nuri
